### ISSUE TITLE
NoImplicitAny ignores inline functions that are type-constrained

### ISSUE THREAD

**Issue by captaincaius** [microsoft/TypeScript#34747]
**TypeScript Version:**  3.8.0-dev.20191025

**Search Terms:**
unknown noimplicitany
noimplicitany inline function

**Code**

```ts
interface SyncProjector<TSource> {
  (source: TSource): any;
}
function wth<T>(func: SyncProjector<T>) {
  return 0;
}
wth(dat=>dat);

```

**Expected behavior:**
The inline dat=>dat function should complain about implicit any

**Actual behavior:**
It doesn't complain even in strict mode.  In fact, I can't figure out any way at all to make it complain about this while constraining func to be some kind of function.

**Playground Link:** 
Already boiled it down to 7 lines above

**Related Issues:** 
None

**Comment by fatcerberus** [comment]
It’s actually not implicit `any`: `dat` gets inferred as *explicit* `any` due to contextual typing.  That in turn causes `T` to be inferred as `any`.

**Comment by captaincaius** [comment]
Why any?  T's type (and hence, TSource's type) isn't specified - vscode and tsplayground say it's "unknown".

Maybe I my example threw you off, since TNext and TSource are in a weird order?  I just removed TNext so it's a little simpler and clearer.

Am I just approaching the problem wrong?  I just want to say "this generic function should have a parameter that needs to be a function with a parameter".

**Comment by fatcerberus** [comment]
Hmm... in the new rewritten example, `dat` is inferred as `unknown`, not `any`.

![image](https://user-images.githubusercontent.com/4097841/67613203-ce40cc00-f778-11e9-8f1d-f743da174231.png)

**Comment by fatcerberus** [comment]
What you'd really need to make this work the way you want, I think, is an "explicit implicit any", then you could go `T = impany` and `dat => dat` would then trigger the error.

**Comment by captaincaius** [comment]
Thank you both for your help.

@fatcerberus yes, you articulated it way better than I could - I understand now why it's behaving the way it does, but the inconsistency between how it behaves when not generic is confusing.

I first thought the "impany" thing you suggested was some hidden feature I didn't know about :D

FYI I tried jimmying it up with what's already available like this:

```typescript
type knownornever<TKnown, TThen> = unknown extends TKnown
  ? // must be unknown or any
  TKnown extends TKnown & string
    ?
      TThen // only any behaves this way
    :
      never // must be unknown
  :
  TThen;

interface SyncProjector<TSource> {
  (source: TSource): any;
}

function wth<T>(func: knownornever<T, SyncProjector<T>>) {
  return 0;
}

wth((dat) => dat); // error \o/ yay
wth((dat: number) => dat); // works \o/ yay
wth((dat: any) => dat); // works \o/ yay

```

It "works" (behaves like a non-generic would), but it's a little hokey and rigid.  If the "impany" thing is possible, that would make this a lot simpler.

**Comment by fatcerberus** [comment]
@captaincaius For posterity, and for anyone else who happens upon this issue in the future, here's a rough play-by-play of what happens:

1. `wth` is called, passing the lambda expression (arrow function) `dat => dat`.
2. `wth<T>` is generic, and we need an inference for `T`.  TS attempts to infer `T` from the parameter type of `func`.  This fails, as `dat` has no type annotation.  There are no other inference sites.
3. In the absence of an inference for `T`, the compiler defaults it.
4. No default is declared for `T`, so its upper bound is chosen--in this case `unknown`.
5. `wth<T>` is now fully instantiated as `wth<unknown>`, which is the final call target.
6. `dat` has no type annotation, so contextual typing kicks in to try to infer it.
7. The `unknown` from type parameter inference is fed back into contextual typing and becomes the type of `dat` (in more prosaic terms, the final call target is not generic; contextual typing infers `func` based on a concrete target type of `(source: unknown) => any`).
8. As a result, no implicit-any error is produced.

**Comment by captaincaius** [comment]
Thanks for the awesome detailed run through!

**Comment by RyanCavanaugh** [comment]
Extremely good summary @fatcerberus 🥇
